zio-http/shared/src/main/scala/zio/http/ServerSentEvent.scala

```diff
@@ -64,23 +62,60 @@ final case class ServerSentEvent(
 }
 
 object ServerSentEvent {
-  implicit lazy val schema: Schema[ServerSentEvent] = DeriveSchema.gen[ServerSentEvent]
 
-  implicit val contentCodec: HttpContentCodec[ServerSentEvent] = HttpContentCodec.from(
-    MediaType.text.`event-stream` -> BinaryCodecWithSchema.fromBinaryCodec(new BinaryCodec[ServerSentEvent] {
-      override def decode(whole: Chunk[Byte]): Either[DecodeError, ServerSentEvent] =
-        throw new UnsupportedOperationException("ServerSentEvent decoding is not yet supported.")
+  implicit def schema[T](implicit schema: Schema[T]): Schema[ServerSentEvent[T]] = DeriveSchema.gen[ServerSentEvent[T]]
+
+  implicit def defaultBinaryCodec[T](implicit schema: Schema[T]): BinaryCodec[ServerSentEvent[T]] =
+    defaultContentCodec(schema).defaultCodec
+
+  private def nextOption(lines: Iterator[String]): Option[String] =
+    if (lines.hasNext) Some(lines.next())
+    else None
+
+  implicit def binaryCodec[T](implicit binaryCodec: BinaryCodec[T]): BinaryCodec[ServerSentEvent[T]] =
+    new BinaryCodec[ServerSentEvent[T]] {
+      override def decode(whole: Chunk[Byte]): Either[DecodeError, ServerSentEvent[T]] = {
+        val lines     = whole.asString.linesIterator
+        val data      = lines.next().stripPrefix("data: ")
+        val eventType = nextOption(lines).map(_.stripPrefix("event: "))
+        val id        = nextOption(lines).map(_.stripPrefix("id: "))
+        val retry     = nextOption(lines).map(_.stripPrefix("retry: ").toInt)
+        val decoded   = binaryCodec.decode(Chunk.fromArray(data.getBytes))
+        decoded.map(value => ServerSentEvent(value, eventType, id, retry))
+      }
 
-      override def streamDecoder: ZPipeline[Any, DecodeError, Byte, ServerSentEvent] =
-        throw new UnsupportedOperationException("ServerSentEvent decoding is not yet supported.")
+      override def streamDecoder: ZPipeline[Any, DecodeError, Byte, ServerSentEvent[T]] =
+        ZPipeline.chunks[Byte].map(_.asString) >>> ZPipeline
+          .splitOn("\n\n")
+          .mapZIO(s => {
+            val lines     = s.linesIterator
```

Comment:
This logic for encoding and decoding should have some tests associated with it. 👍

Reply:
@jdegoes This is actually more something for #2953 since receiving SSE is not yet impl. I have a draft PR open for that #2987 that I am working on right now. Are you okay if I add the tests there?